rfd/0080-yubikey-piv.md

```diff
@@ -0,0 +1,468 @@
+
+The user can provide `--piv-pin-policy=<never|always|once>` and `--piv-touch-policy=<never|always|cached>` to specify their desired policies.
+
+#### PIV slots
+
+PIV cards are required to have 4 primary slots, as well as the 20 retired key management slots. 
+
+Each slot has a different intended function and different default PIN and touch policies. Since we are manually setting the PIN and touch policies ourselves, we can technically use any of the 24 slots available in the same way.
+
+We will use slot `9a: PIV Authentication` by default, as it fits the teleport use case most closely. However, users will be able to provide their own desired slot with the flag `--piv-slot`, in case they are already using slot `9a` for a different login session or for a different PIV application, such as `yubikey-agent`.
+
+Note: Yubikeys also have an attestation slot, `f9`, which is not standardized by the PIV specifications.
+
+#### tsh PIV login session
+
+Usually, `tsh` checks the user's `~/.tsh/keys/proxy.example.com` for an existing public-private key pair. If found, that key pair will be used alongside any stored certificates to handle `tsh` calls.
+
+Since we can't store the PIV private key on disk, we will instead store a fake private key which we can use to identify the PIV slot to load the private key from. This key will be PEM encoded data where, `pem.Type` specifies the PIV key type, and the encoded data contains a unique ID and PIV slot.
+
+Note: for yubikey, the unique ID will be the card's serial number. Different implementations of PIV may provide alternative methods for identifying unique cards
+
+```
+-----BEGIN PIV YUBIKEY PRIVATE KEY-----
+# PEM encoded `yubikey_serial_number+piv_slot`
+-----END PIV YUBIKEY PRIVATE KEY-----
+```
+
+##### Check the public key
+
+It's possible that after login, the user overwrites the PIV slot used to login. To catch this, future calls to `tsh` will retrieve the public key from the PIV slot and check if it matches the public key on the user's TLS certificate (`~/.tsh/keys/proxy/user-x509.pem`). If it doesn't match, then the user will see an error and get prompted to relogin.
+
+```
+> tsh ls
+ERROR: PIV private key does not match login certificates.
+// Trigger relogin
+```
+
+##### Concurrent login sessions
+
+`tsh` should support multiple concurrent login sessions for different proxies. This can be done by specifying a different slot for each concurrent login session, allowing up to 24 concurrent PIV login sessions. To make it easier for `tsh` to track used slots, and avoid overwriting slots that are already in use, `tsh login --piv` will also create a self-signed certificate with login metadata attached and add it to the slot.
+
+```
+Slot 9a:
+    Certificate:
+        Subject DN:	CN = yubikey, O = tsh, OU = v10.0.0, L = proxy.example.com:user
+```
+
+Now, future calls to `tsh login --piv --piv-slot=9a` can retrieve the certificate from the slot and decide what to do:
```

Comment:
nit: I'm not sure if this is a good idea, but I think that if `--piv-slot` or any other piv-related switch are provided, then `--piv` shouldn't be required. Otherwise, we would probably need to display an error message that --piv is missing.